### PR TITLE
Authentication

### PR DESCRIPTION
**Pull request by LetsTrie**

### CODE REVIEW

File: config/logger.js [hunk 1]
```diff
@@ -0,0 +1,23 @@
...
+const winston = require('winston');
+
+module.exports = createLogger({
+  level: 'info',
+  format: format.combine(
+    format.simple(),
+    format.timestamp(),
+    format.printf(info => `[${info.timestamp}] ${info.level}: ${info.message}`)
+  ),
+  defaultMeta: { service: 'user-service' },
```
Comment (orvi2014) [inline config/logger.js]: why do you need this line?

File: config/logger.js [hunk 2]
```diff
@@ -0,0 +1,23 @@
+const { createLogger, format, transports } = require('winston');
+const winston = require('winston');
+
+module.exports = createLogger({
+  level: 'info',
+  format: format.combine(
+    format.simple(),
+    format.timestamp(),
+    format.printf(info => `[${info.timestamp}] ${info.level}: ${info.message}`)
```
Comment (orvi2014) [inline config/logger.js]: Add `file name` and `line number` to improve the readability of the log

File: .gitignore [hunk 3]
```diff
@@ -0,0 +1,107 @@
...
+.fusebox/
+
+# DynamoDB Local files
+.dynamodb/
+
+# TernJS port file
+.tern-port
+
+# Todo list
+todo.txt
```
Comment (orvi2014) [inline .gitignore]: add `package-lock.json` here

File: LICENSE [hunk 4]
```diff
@@ -0,0 +1,21 @@
+MIT License
+
+Copyright (c) 2020 MD. Sakib Khan
```
Comment (orvi2014) [inline LICENSE]: Change the `copyright` issue

File: README.md [hunk 5]
```diff
@@ -1 +1 @@
-# Backend-Modules
\ No newline at end of file
+# enedao
```
Comment (orvi2014) [inline README.md]: It should be `backend-modules`

File: server/auth/auth.controller.js [hunk 6]
```diff
@@ -0,0 +1,247 @@
...
+  user.verifyAccountToken = undefined;
+  user.isAccountVerified = true;
+  await user.save();
+  return res.status(200).json({ success: true });
+});
+
+exports.forgetPasswordSendToken = asyncHandler(async (req, res, next) => {
+  const { error } = forgetPasswordVal(req.body);
+  if (error) return next(new ErrorResponse(error.details[0].message, 400));
+  const resetToken = crypto.randomBytes(20).toString('hex');
```
Comment (orvi2014) [inline server/auth/auth.controller.js:106]: can you explain this part?

File: server/auth/auth.controller.js [hunk 7]
```diff
@@ -0,0 +1,247 @@
...
+    return res.status(200).json({ success: true });
+  }
+  return next(new ErrorResponse('Old password is not matching', 401));
+});
+
+exports.updateDetails = asyncHandler(async (req, res, next) => {
+  let { username, email, phoneNumber } = req.body;
+  if (!username) username = req.user.username;
+  if (!email) email = req.user.email;
+  if (!phoneNumber) phoneNumber = req.user.phoneNumber;
```
Comment (orvi2014) [inline server/auth/auth.controller.js:179]: Explain this line. Did you take whole user info into `req.user`?

File: server/auth/auth.controller.js [hunk 8]
```diff
@@ -0,0 +1,247 @@
...
+
+  const fieldsToUpdate = { username, email, phoneNumber };
+
+  const user = await AuthModel.findByIdAndUpdate(req.user._id, fieldsToUpdate, {
+    new: true
```
Comment (orvi2014) [inline server/auth/auth.controller.js:184]: does it create a new document or update the previous document?

File: server/auth/auth.controller.js [hunk 9]
```diff
@@ -0,0 +1,247 @@
...
+  const refreshToken = req.body.refreshToken;
+
+  try {
+    const isBlackListed = await redis.get(`BlackListed${refreshToken}`);
+    if (isBlackListed) {
+      return next(new ErrorResponse('This Token is already used', 403));
+    }
+    const decoded = await jwt.verify(refreshToken, config.jwtRefreshKey);
+    const user = await AuthModel.findById(decoded.id);
+    if (!user) return next(new ErrorResponse('User not found', 404));
```
Comment (orvi2014) [inline server/auth/auth.controller.js]: Do you need these lines of code? I don't think so. why do I need to know if a user exist or not during refresh my token?

File: server/auth/auth.controller.js [hunk 10]
```diff
@@ -0,0 +1,247 @@
...
+
+exports.deleteAccount = asyncHandler(async (req, res, next) => {
+  await AuthModel.findByIdAndDelete({ _id: req.user._id });
+  req.user = undefined;
+  return res.status(200).json({ success: true });
+});
+
+exports.tokenRefresher = asyncHandler(async (req, res, next) => {
+  const refreshToken = req.body.refreshToken;
+
```
Comment (orvi2014) [inline server/auth/auth.controller.js:206]: Don't you need to take `access token`?

File: server/auth/auth.controller.js [hunk 11]
```diff
@@ -0,0 +1,247 @@
...
+  }
+});
+
+exports.logout = asyncHandler(async (req, res, next) => {
+  const { refreshToken } = req.body;
+  const { accessToken } = req;
+  if (!refreshToken)
+    return next(new ErrorResponse('Refresh Token not found!! ', 404));
+
+  await redis.set(`BlackListed${refreshToken}`, accessToken, 'EX', 30 * 60);
```
Comment (orvi2014) [inline server/auth/auth.controller.js]: Why do you set token expiration manually? you have to set the valid time against current current time.

File: server/middleware/authorization.js [hunk 12]
```diff
@@ -0,0 +1,33 @@
...
+  const bearerHeader = req.headers.authorization;
+  if (bearerHeader && bearerHeader.startsWith('Bearer')) {
+    const accessToken = bearerHeader.split(' ')[1];
+
+    if (!accessToken) return next(new ErrorResponse('Not authorized !!', 401));
+
+    try {
+      const isBlackListed = await redis.get(`BlackListed${accessToken}`);
+      if (isBlackListed) {
+        return next(new ErrorResponse('This Token is already used', 403));
```
Comment (orvi2014) [inline server/middleware/authorization.js]: You should say 'the token is invalid' and check the status code.. Is it `403` or `401`

File: server/middleware/authorization.js [hunk 13]
```diff
@@ -0,0 +1,33 @@
...
+      }
+      const decoded = await jwt.verify(accessToken, config.jwtAccessKey);
+      const user = await AuthModel.findById(decoded.id);
+      if (!user) return next(new ErrorResponse('User not found', 404));
+
+      req.accessToken = accessToken;
```
Comment (orvi2014) [inline server/middleware/authorization.js]: why do I need this line?

File: config/logger.js (hunk 1 above)
Comment (LetsTrie) [reply config/logger.js]: defaultMeta !!! I don't know. It is in the documentation.

File: config/logger.js (hunk 2 above)
Comment (LetsTrie) [reply config/logger.js]: I found file name in "module". But couldn't figure out line number. need help

Comment (LetsTrie) [reply config/logger.js]: In which which place, I should place logger? in API?

File: .gitignore (hunk 3 above)
Comment (LetsTrie) [reply .gitignore]: Done. 
Thanks. I should've add this here from the beginning :+1:

File: LICENSE (hunk 4 above)
Comment (LetsTrie) [reply LICENSE]: I selected Licence term while creating repo :P I dont know much about it.
deleted.

File: README.md (hunk 5 above)
Comment (LetsTrie) [reply README.md]: Done

File: server/auth/auth.controller.js (hunk 6 above)
Comment (LetsTrie) [reply server/auth/auth.controller.js:106]: Random crypto token !!  https://nodejs.org/api/crypto.html#crypto_crypto_randombytes_size_callback

File: server/auth/auth.controller.js (hunk 7 above)
Comment (LetsTrie) [reply server/auth/auth.controller.js:179]: from my code, everything except password

File: server/auth/auth.controller.js (hunk 8 above)
Comment (LetsTrie) [reply server/auth/auth.controller.js:184]: After updating, I was not getting the correct updated model. 
then after searching, I found this => https://stackoverflow.com/a/30419860/9818882
By adding this, It returns the current updated model

File: server/auth/auth.controller.js (hunk 9 above)
Comment (LetsTrie) [reply server/auth/auth.controller.js]: ok. Changed

File: server/auth/auth.controller.js (hunk 10 above)
Comment (LetsTrie) [reply server/auth/auth.controller.js:206]: Should I check access token is expired? Then I will give refresh token ?

File: server/middleware/authorization.js (hunk 12 above)
Comment (LetsTrie) [reply server/middleware/authorization.js]: Thank you. Changed.

File: server/auth/auth.controller.js (hunk 10 above)
Comment (LetsTrie) [reply server/auth/auth.controller.js:206]: or Should I check the expired access token is already is in the blacklist???

File: server/middleware/authorization.js (hunk 13 above)
Comment (LetsTrie) [reply server/middleware/authorization.js]: deleted.

**Pull request merged**